### PR TITLE
Rotate to angle command using PID controller

### PR DESCRIPTION
**Pull request by jherrmann06**

### CODE REVIEW

File: src/main/java/frc/robot/commands/RotateCommand.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            return true;
+        } 
+
+        return false;
+    }
+
+    @Override
+    public void execute() {
+        SmartDashboard.putNumber("radians away", drive.getGyroscopeRotation().minus(target).getRadians());
+        drive.rotate(rotationDirection, pid.calculate(Math.abs(drive.getGyroscopeRotation().minus(target).getRadians())) * -1);
```
Comment (jdlanning) [inline src/main/java/frc/robot/commands/RotateCommand.java]: Create local variable for the rotation error (diff) and only call once

File: src/main/java/frc/robot/Constants.java [hunk 2]
```diff
@@ -1,10 +1,12 @@
 package frc.robot;
 
 public final class Constants {
-    public static final int DRIVER_CONTROLLER = 0;
     public static final double DRIVETRAIN_TRACKWIDTH_METERS = 0.5969; //Measured on robot
     public static final double DRIVETRAIN_WHEELBASE_METERS = 0.5969;
 
+    public static final double DEFAULT_RADIANS_PER_SECOND = 3.0;
+    public static final double ACCEPTABLE_AUTOROTATE_ERROR_DEGREES = 3.0;
```
Comment (jdlanning) [inline src/main/java/frc/robot/Constants.java]: DRIVE_ROTATE_TOLERANCE is a more clear name, IMO

File: src/main/java/frc/robot/commands/RotateCommand.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.Constants;
+import frc.robot.subsystems.DrivetrainSubsystem;
+
+public class RotateCommand extends CommandBase{
+    private DrivetrainSubsystem drive;
+    private Rotation2d target;
+    private boolean rotationDirection;
+
+    private PIDController pid = new PIDController(3, 0.5, 0.5);
```
Comment (jdlanning) [inline src/main/java/frc/robot/commands/RotateCommand.java]: These would be good to move to constants

File: src/main/java/frc/robot/commands/RotateCommand.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            rotationDirection = false;
+        }
+    }
+
+    @Override
+    public boolean isFinished() {
+        return withinAcceptableOffset();
+    }
+
+    private boolean withinAcceptableOffset() {
```
Comment (jdlanning) [inline src/main/java/frc/robot/commands/RotateCommand.java]: withinTolerance to align with my other naming suggestion

File: src/main/java/frc/robot/commands/RotateCommand.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        if(drive.getGyroscopeRotation().equals(target)) {
+            return true;
+        }
```
Comment (jdlanning) [inline src/main/java/frc/robot/commands/RotateCommand.java]: This if statement is not needed as below will always be true when it is true

File: src/main/java/frc/robot/commands/RotateCommand.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    public void execute() {
+        SmartDashboard.putNumber("radians away", drive.getGyroscopeRotation().minus(target).getRadians());
+        drive.rotate(rotationDirection, pid.calculate(Math.abs(drive.getGyroscopeRotation().minus(target).getRadians())) * -1);
+    }
+
+    @Override 
+    public void end(boolean interrupted) {
+        drive.drive(new ChassisSpeeds(0, 0, 0));
```
Comment (jdlanning) [inline src/main/java/frc/robot/commands/RotateCommand.java:41]: Nice and safe end to the command !

File: src/main/java/frc/robot/commands/RotateCommand.java (hunk 1 above)
Comment (jdlanning) [inline src/main/java/frc/robot/commands/RotateCommand.java]: Ideally we should not need to know the direction and the rotation direction is based on the sign of the target error. Therefore no need to take abs of the value as the sign represents the direction

File: src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java [hunk 7]
```diff
@@ -150,4 +151,17 @@ public void periodic() {
...
+  public void rotate(boolean directionIsRight, double radiansPerSecond) {
+          if(directionIsRight) {
+                  drive(new ChassisSpeeds(0, 0, -radiansPerSecond));
+          }else {
+                drive(new ChassisSpeeds(0, 0, radiansPerSecond));
+          }
+  }
+
+  public void rotate(boolean directionIsRight) {
+        rotate(directionIsRight, Constants.DEFAULT_RADIANS_PER_SECOND);
```
Comment (jdlanning) [inline src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java]: This method is unused and can be removed

**Review by jdlanning — COMMENTED** [review]
Nice work ! Super impressed how quickly and independently you all got a PID rotate command up and running

**Comment by jherrmann06** [comment]
Responded to feedback

**Review by jdlanning — APPROVED** [review]

**Pull request merged**